File: server/src/main/java/com/solid0us/time_quest_log/repositories/UserGameRepository.java
package com.solid0us.time_quest_log.repositories;

import com.solid0us.time_quest_log.model.DTOs.UserGameWithHoursDTO;
import com.solid0us.time_quest_log.model.Games;
import com.solid0us.time_quest_log.model.UserGames;
import com.solid0us.time_quest_log.model.Users;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface UserGameRepository extends JpaRepository<UserGames, UUID> {
    List<UserGames> findByUser_Id(UUID userId);

    Optional<UserGames> findByUserAndGame(Users user, Games game);


    @Query(value = """
           SELECT COUNT(DISTINCT game_id)
           FROM user_games
           WHERE user_id = :userId;
    """, nativeQuery = true)
    int findDistinctGamesByUserId(@Param("userId") UUID userId);

    @Query(value = """
       SELECT
           g.id AS gameId,
           g.name AS gameName,
           g.cover_url AS coverUrl,
           ug.exe_name AS exeName,
           genres.genre_list AS genres,
           COALESCE(hours.total_hours, 0) AS hoursPlayed
       FROM
           user_games ug
       JOIN
           games g ON g.id = ug.game_id
       JOIN
           users u ON u.id = ug.user_id
       LEFT JOIN (
           SELECT
               game_id,
               SUM(EXTRACT(EPOCH FROM (end_time - start_time)) / 3600)::double precision AS total_hours
           FROM
               game_sessions
           WHERE
              user_id = :userId
           GROUP BY
               game_id
       ) hours ON hours.game_id = g.id
       LEFT JOIN (
           SELECT
               g_gen.game_id,
               STRING_AGG(DISTINCT gen.name, ', ') AS genre_list
           FROM
               game_genre g_gen
           JOIN
               genres gen ON gen.id = g_gen.genre_id
           GROUP BY
               g_gen.game_id
       ) genres ON genres.game_id = g.id
       WHERE
           u.id = :userId
       ORDER BY g.name
""", nativeQuery = true)
    List<UserGameWithHoursDTO> getUserGameWithHoursPlayed(@Param("userId") UUID userId);
}

File: TimeQuestLogDesktopApp/TimeQuestLogDesktopApp/Repositories/UserGameRepository.cs
using Dapper;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TimeQuestLogDesktopApp.Database;
using TimeQuestLogDesktopApp.Models;
using TimeQuestLogDesktopApp.Models.DTOs;

namespace TimeQuestLogDesktopApp.Repositories
{
    internal class UserGameRepository : SqliteDataAccess
	{
		private readonly IDbConnectionFactory _connectionFactory;

        public UserGameRepository(IDbConnectionFactory connectionFactory)
        {
            _connectionFactory = connectionFactory; 
        }

        public List<UserGameDTO> GetUserGames(string userId)
        {
            List<UserGameDTO> userGames = new List<UserGameDTO>();
			using (var cnn = _connectionFactory.CreateConnection())
			{
				string sql = @"
                SELECT 
					game.id,
                    game.Name, 
                    game.CoverUrl,
                    u.Username,
                    ug.ExeName,
                    gen.Id,
                    gen.Name
                FROM UserGames ug
                INNER JOIN Users u ON u.Id = ug.UserId
                INNER JOIN Games game ON game.Id = ug.GameId
                LEFT JOIN GameGenre gg ON game.Id = gg.GameId
                LEFT JOIN Genres gen ON gg.GenreId = gen.Id
                WHERE u.Id = @UserId
                ORDER BY game.Name";
				var gamesDict = new Dictionary<int, UserGameDTO>();

				cnn.Query<Games, Users, UserGames, Genres, UserGameDTO>(
					sql,
					(game, user, userGame, genre) =>
					{
						if (!gamesDict.TryGetValue(game.Id, out UserGameDTO existingGame))
						{
							UserGameDTO newUserGame = new UserGameDTO(game.Id, game.Name, game.CoverUrl, userGame.ExeName, user.Username);

							gamesDict.Add(game.Id, newUserGame);
							existingGame = newUserGame;
						}

						if (genre != null && !existingGame.Genres.Any(g => g.Id == genre.Id))
						{
							existingGame.Genres.Add(genre);
						}

						return existingGame;
					},
					param: new { UserId = userId },
					splitOn: "Username,ExeName,Id"
				);

				return gamesDict.Values.ToList();
			}
        }

		public int CreateUserGame(UserGames userGame)
		{
			using(var cnn = _connectionFactory.CreateConnection())
			{
				string sql = @"
					INSERT OR IGNORE INTO UserGames
						(Id, ExeName, GameId, UserId)
					VALUES	
						(@Id, @ExeName, @GameId, @UserId)
				";
				var parameters = new { userGame.Id, userGame.ExeName, userGame.GameId, userGame.UserId };

				return cnn.Execute(sql, parameters);
			}
		}

		public int UpdateSync(string id, bool isSynced)
		{
			using(var cnn = _connectionFactory.CreateConnection())
			{
				int syncValue = isSynced ? 1 : 0;
				string sql = @"
					UPDATE UserGames
					SET IsSynced = @SyncValue
					WHERE Id = @Id
				";
				var parameters = new { Id = id, SyncValue = syncValue };

				return cnn.Execute(sql, parameters);
			}
		}

		public IEnumerable<UserGameDTO> GetUnsyncedUserGamesByUserId(string userId)
		{
			using(var cnn = _connectionFactory.CreateConnection())
			{
				string sql = @"
                SELECT 
					game.id,
                    game.Name, 
                    game.CoverUrl,
                    u.Username,
					ug.Id,
                    ug.ExeName,
                    gen.Id,
                    gen.Name
                FROM UserGames ug
                INNER JOIN Users u ON u.Id = ug.UserId
                INNER JOIN Games game ON game.Id = ug.GameId
                LEFT JOIN GameGenre gg ON game.Id = gg.GameId
                LEFT JOIN Genres gen ON gg.GenreId = gen.Id
                WHERE u.Id = @UserId
				AND ug.IsSynced = 0";
				var gamesDict = new Dictionary<int, UserGameDTO>();

				cnn.Query<Games, Users, UserGames, Genres, UserGameDTO>(
					sql,
					(game, user, userGame, genre) =>
					{
						if (!gamesDict.TryGetValue(game.Id, out UserGameDTO existingGame))
						{
							UserGameDTO newUserGame = new UserGameDTO(userGame.Id, game.Id, game.Name, game.CoverUrl, userGame.ExeName, user.Username);

							gamesDict.Add(game.Id, newUserGame);
							existingGame = newUserGame;
						}

						if (genre != null && !existingGame.Genres.Any(g => g.Id == genre.Id))
						{
							existingGame.Genres.Add(genre);
						}

						return existingGame;
					},
					param: new { UserId = userId },
					splitOn: "Username,Id,Id"
				);

				return gamesDict.Values.ToList();
			}
		}
    }
}
